internal/provider/datasources/block.go

```diff
@@ -0,0 +1,198 @@
+}
+
+// Schema defines the scema for the data source.
+func (d *blockDataSource) Schema(_ context.Context, _ datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: `
+Get information about an existing Block by either:
+- ID, or
+- block type name and block name
+<br>
+If the ID is provided, then the block type name and block name will be ignored.
+<br>
+Use this data source to obtain Block-specific attributes, such as the data.
+`,
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:    true,
+				CustomType:  customtypes.UUIDType{},
+				Description: "Block ID (UUID)",
+				Optional:    true,
+			},
+			"created": schema.StringAttribute{
+				Computed:    true,
+				CustomType:  customtypes.TimestampType{},
+				Description: "Timestamp of when the resource was created (RFC3339)",
+			},
+			"updated": schema.StringAttribute{
+				Computed:    true,
+				CustomType:  customtypes.TimestampType{},
+				Description: "Timestamp of when the resource was updated (RFC3339)",
+			},
+			"account_id": schema.StringAttribute{
+				CustomType:  customtypes.UUIDType{},
+				Description: "Account ID (UUID), defaults to the account set in the provider",
+				Optional:    true,
+			},
+			"workspace_id": schema.StringAttribute{
+				CustomType:  customtypes.UUIDType{},
+				Description: "Workspace ID (UUID), defaults to the workspace set in the provider",
+				Optional:    true,
+			},
+			"name": schema.StringAttribute{
+				Computed:    true,
+				Description: "Name of the block",
+				Optional:    true,
+			},
+			"data": schema.StringAttribute{
+				Computed:    true,
+				Sensitive:   true,
+				CustomType:  jsontypes.NormalizedType{},
+				Description: "The user-inputted Block payload, as a JSON string. The value's schema will depend on the selected `type` slug. Use `prefect block types inspect <slug>` to view the data schema for a given Block type.",
+			},
+			"block_type_name": schema.StringAttribute{
+				Computed:    true,
+				Description: "Block type name",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+// Read refreshes the Terraform state with the latest data.
+func (d *blockDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state BlockDataSourceModel
+
+	diag := req.Config.Get(ctx, &state)
+	resp.Diagnostics.Append(diag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client, err := d.client.BlockDocuments(state.AccountID.ValueUUID(), state.WorkspaceID.ValueUUID())
+	if err != nil {
+		resp.Diagnostics.Append(helpers.CreateClientErrorDiagnostic("Block", err))
+
+		return
+	}
+
+	var block *api.BlockDocument
+
+	switch {
+	case !state.ID.IsNull():
+		block, err = client.Get(ctx, state.ID.ValueUUID())
+	case !state.Name.IsNull() && !state.BlockTypeName.IsNull():
+		block, err = client.GetByName(ctx, state.BlockTypeName.ValueString(), state.Name.ValueString())
```

Comment:
yeah now that you mention it - i might actually modify the `BlockTypeName / block_type_name` to be `TypeSlug / type_slug`

the API expects a slug

```
@router.get(
    "/slug/{slug}/block_documents/name/{block_document_name}",
)
```

Reply:
I'm down, the one caveat I see is that the [API docs](https://app.prefect.cloud/api/docs#tag/Block-documents/operation/read_block_document_by_id_api_accounts__account_id__workspaces__workspace_id__block_documents__id__get) specifically call it `block_type_name` 🤔

Reply:
Mm, I see what you mean. I was mapping the [BlockDocument class from Nebula](https://github.com/PrefectHQ/nebula/blob/22e66a334369d96a08fc19a877c9117e94f388b8/src/prefect/server/schemas/core.py#L681-L704) to the [BlockDocument](https://github.com/PrefectHQ/terraform-provider-prefect/blob/b1f0f16ebd55d272d7f8aca29b76038b8576ad4d/internal/api/block_documents.go#L20-L31) struct, which is where I got `block_type_name`. Looks like I just need to mentally delinate when objects represent the API request vs the API response.

Reply:
Iterated on this in 95ae8f5, lemme know what you think.